### PR TITLE
Devrel 1068 import item and variants

### PR DESCRIPTION
**Pull request by JiriLojda**
### Motivation

Which issue does this fix? Fixes #`issue number`

If no issue exists, what is the fix or new feature? Were there any reasons to fix/implement things that are not obvious?

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

If manual testing is required, what are the steps?

### CODE REVIEW

**Comment by IvanKiral** [comment]
let's test the situation when having option of multiple choice element with name same as some other element in the same content_type.. external_ids might collide

**Review by IvanKiral — CHANGES_REQUESTED** [review]
In general, there might be problem with external ids. Otherwise the logic looks good, however, I would do some refactoring, code is hard to follow, I would extract some variables of functions

File: src/commands/importExportEntities/entities/contentTypesSnippets.ts [hunk 1]
```diff
@@ -21,15 +21,42 @@ export const contentTypesSnippetsEntity: EntityDefinition<
...
           .map(([fileSnippet, projectSnippet]) => {
             const elementIdEntries = zip(fileSnippet.elements, projectSnippet.elements)
               .map(([fileEl, projectEl]) => [fileEl.id ?? "", projectEl.id ?? ""] as const);
 
-            return [fileSnippet.id, { selfId: projectSnippet.id, elementIdsByOldIds: new Map(elementIdEntries) }];
+            return [fileSnippet.id, {
+              selfId: projectSnippet.id,
+              elementIdsByOldIds: new Map(elementIdEntries),
+              elementTypeByOldIds: new Map(fileSnippet.elements.map(el => [el.id ?? "", el.type])),
+              multiChoiceOptionIdsByOldIdsByOldElementId: new Map(
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/commands/importExportEntities/entities/contentTypesSnippets.ts:34]: I refactored the function, I believe it is no longer necessary to extract it. What do you think?

File: src/commands/importExportEntities/entities/languageVariants.ts [hunk 2]
```diff
@@ -15,18 +23,300 @@ export const languageVariantsExportEntity: EntityDefinition<
...
+    if (!typeContext) {
+      throw new Error(`Cannot find type context for item "${fileVariant.item.id}".`);
+    }
+
+    const newWfContext = findTargetWfStep(context, fileVariant);
+
+    const projectVariant = await client
+      .upsertLanguageVariant()
+      .byItemId(context.contentItemContextByOldIds.get(fileVariant.item.id ?? "")?.selfId ?? "")
+      .byLanguageId(context.languageIdsByOldIds.get(fileVariant.language.id ?? "") ?? "")
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/commands/importExportEntities/entities/languageVariants.ts:56]: Well, the best thing to do would be to improve the types. The next best thing is to throw an exception and then probably the `as` cast. 😄 We use this approach in many places across the data-ops codebase (e.g. a variant also always has an item as well... I'd like to solve this systematically instead of fixing this one place. I am thinking about creating a proper type and casting to it (e.g. `as BetterReference`) and using it across the codebase. Then, once we fix the SDK, we can just delete the type and have it fixed across the whole solution once again. What do you think?

File: src/commands/importExportEntities/entities/languageVariants.ts [hunk 3]
```diff
@@ -15,18 +23,300 @@ export const languageVariantsExportEntity: EntityDefinition<
...
+      .withData(builder => ({
+        workflow: {
+          workflow_identifier: {
+            id: newWfContext.wfId,
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/commands/importExportEntities/entities/languageVariants.ts:60]: Yeah, but isn't wf an obvious abbreviation in this context?

File: src/commands/importExportEntities/entities/languageVariants.ts [hunk 4]
```diff
@@ -15,18 +23,300 @@ export const languageVariantsExportEntity: EntityDefinition<
...
+  variant: LanguageVariantContracts.ILanguageVariantModelContract,
+  scheduleTo?: Date,
+) => {
+  const sharedRequest = client
+    .publishLanguageVariant()
+    .byItemId(variant.item.id ?? "")
+    .byLanguageId(variant.language.id ?? "");
+
+  return scheduleTo
+    ? await sharedRequest
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/commands/importExportEntities/entities/languageVariants.ts:110]: I are right, I do not. 👍

File: src/commands/importExportEntities/entities/languageVariants.ts [hunk 5]
```diff
@@ -15,18 +23,300 @@ export const languageVariantsExportEntity: EntityDefinition<
...
+  elementTypeByOldId: ReadonlyMap<string, string>;
+  elementIdByOldId: ReadonlyMap<string, string>;
+  multiChoiceOptionIdsByOldIdsByOldElementId: ReadonlyMap<string, ReadonlyMap<string, string>>;
+}>;
+
+const createTransformElement = (params: TransformElementParams) =>
+(
+  fileElement: ElementContracts.IContentItemElementContract,
+): LanguageVariantElements.ILanguageVariantElementBase | null => {
+  const elementType = params.elementTypeByOldId.get(fileElement.element.id ?? "");
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/commands/importExportEntities/entities/languageVariants.ts:132]: See the other comment I made.

File: src/commands/importExportEntities/entities/utils/typeElements.ts [hunk 6]
```diff
@@ -272,7 +273,7 @@ export const createPatchItemAndTypeReferencesInTypeElement =
             op: "replace" as const,
             path: `/elements/id:${newElementId}/allowed_content_types`,
             value: typedElement.allowed_content_types.map(ref => ({
-              id: context.contentTypeIdsWithElementsByOldIds.get(ref.id ?? "")?.selfId,
+              id: context.contentTypeContextWithElementsByOldIds.get(ref.id ?? "")?.selfId,
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/commands/importExportEntities/entities/utils/typeElements.ts:276]: The codenames are not in scope. I would have to add them into the context, but the ids are already there so I can as well use them.

File: src/commands/importExportEntities/entities/languageVariants.ts [hunk 7]
```diff
@@ -15,18 +23,300 @@ export const languageVariantsExportEntity: EntityDefinition<
...
+      };
+    }
+  }
+};
+
+const createTranslateWfStepId = (context: ImportContext) => (stepId: string): string =>
+  context.worfklowStepsIdsWithTransitionsByOldIds.get(stepId)?.selfId ?? "";
+
+const getProjectElementId = (params: TransformElementParams, fileId: string): string | null =>
+  Array.from(params.context.contentTypeSnippetContextWithElementsByOldIds.values())
```
[the remark this answers is not in the document]
Comment (JiriLojda) [reply src/commands/importExportEntities/entities/languageVariants.ts]: I managed to inline the snippets here as well. The snippet elements don't need an update later on so I don't need their ids anyway.

**Comment by JiriLojda** [comment]
> let's test the situation when having option of multiple choice element with name same as some other element in the same content_type.. external_ids might collide

It seems that the external id of a multi-choice element's option must be unique even across multiple types. I changed it to use external ids like so: `{type/snippet codename}_{element codename}_{option codename}`

**Comment by JiriLojda** [comment]
Since Ivan is on a long vacation, I will merge this for now. We can talk about the comments later.
